pkg/controller/acmeorders/sync.go

```diff
@@ -380,6 +389,157 @@ func (c *Controller) createOrder(ctx context.Context, cl acmecl.Interface, issue
 }
 
 func (c *Controller) challengeSpecForAuthorization(ctx context.Context, cl acmecl.Interface, issuer cmapi.GenericIssuer, o *cmapi.Order, authz *acmeapi.Authorization) (*cmapi.ChallengeSpec, error) {
+	// 1. fetch solvers from issuer
+	solvers := issuer.GetSpec().ACME.Solvers
+
+	// 2. filter solvers to only those that matchLabels
+	var candidates []cmapi.ACMEChallengeSolver
+	for _, cfg := range solvers {
+		// if this config has no selector at all, then it can be used for all
+		// domain authorizations, so we include it
+		if cfg.Selector == nil {
+			candidates = append(candidates, cfg)
+			continue
+		}
+		if !resourceMatchesSelector(o, cfg.Selector.MatchLabels) {
+			continue
+		}
+		if len(cfg.Selector.DNSNames) > 0 && !orderHasOneOfDNSNames(o, cfg.Selector.DNSNames...) {
+			continue
+		}
+		candidates = append(candidates, cfg)
+	}
+
+	challengeForSolver := func(solver *cmapi.ACMEChallengeSolver) *acmeapi.Challenge {
+		for _, ch := range authz.Challenges {
+			switch {
+			case ch.Type == "http-01" && solver.HTTP01 != nil:
+				return ch
+			case ch.Type == "dns-01" && solver.DNS01 != nil:
+				return ch
+			}
+		}
+		return nil
+	}
+
+	// 3. iterate through each solver, finding the most specific match (taking account of dnsNames)
+	domainToFind := authz.Identifier.Value
+	if authz.Wildcard {
+		domainToFind = "*." + domainToFind
+	}
+
+	// if a solver config that matches all dns names is found, we'll use the
+	// one with the most labels, as this is the 'most specific match' for the
+	// certificate this order is fulfilling.
+	// the matchAll solver is only used *iff* the domainToFind is not listed in
+	// any other solver's DNSNames list.
+	acmeCh, solverConfigToUse := func() (*acmeapi.Challenge, *cmapi.ACMEChallengeSolver) {
```

Comment:
Why is this an anonymous function? It seems like it would be more explicit to make it a standalone function that takes `candidates`, `challengeForSolver` and `domainToFind` as parameters. Not a blocker on merging though